File: kristian/solutions/martinvl.java
/*
@EXPECTED_RESULTS@: CORRECT
*/

import java.util.Scanner;
import java.util.LinkedList;
import java.util.Arrays;

public class martinvl {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int N = in.nextInt();
        int M = in.nextInt();
        in.nextLine();

        boolean[][] grid = new boolean[M][N];
        int[] start = new int[]{0, 0};
        int[] goal = new int[]{0, 0};

        for (int i = 0; i < M; ++i) {
            char[] part = in.nextLine().toCharArray();

            for (int j = 0; j < N; ++j) {
                grid[i][j] = part[j] != 'X';

                if (part[j] == 'S') {
                    start = new int[]{i, j};
                } else if (part[j] == 'G') {
                    goal = new int[]{i, j};
                }
            }
        }

        int dist = bfs(grid, start, goal);

        if (dist == -1) {
            System.out.println("STAY WHERE YOU ARE");
        } else {
            System.out.println(dist-1);
        }
    }

    static final int[][] DP = new int[][]{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    static int bfs(boolean[][] grid, int[] start, int[] goal) {
        int M = grid.length;
        int N = grid[0].length;
        int[][] dist = new int[M][N];
        LinkedList<int[]> q = new LinkedList<int[]>();

        for (int i = 0; i < M; ++i) {
            Arrays.fill(dist[i], -1);
        }

        dist[start[0]][start[1]] = 0;
        q.add(start);

        while (!q.isEmpty()) {
            int[] p = q.poll();

            if (p[0] == goal[0] && p[1] == goal[1]) {
                break;
            }

            for (int[] dp : DP) {
                int[] np = new int[]{p[0] + dp[0], p[1] + dp[1]};

                if (np[0] < 0 || np[0] >= M || np[1] < 0 || np[1] >= N) {
                    continue;
                }

                if (dist[np[0]][np[1]] != -1 || !grid[np[0]][np[1]]) {
                    continue;
                }

                dist[np[0]][np[1]] = dist[p[0]][p[1]] + 1;
                q.add(np);
            }
        }

        return dist[goal[0]][goal[1]];
    }
}

File: kristian/solutions/martinvl.py
# @EXPECTED_RESULTS@: CORRECT

from sys import stdin

N, M = map(int, stdin.next().split())
raw_grid = ''.join(stdin).replace('\n', '')

# find start and goal
start_idx = raw_grid.index('S')
start = (start_idx / N, start_idx % N)

goal_idx = raw_grid.index('G')
goal = (goal_idx / N, goal_idx % N)

# find open intersections
opened = [[c != 'X' for c in raw_grid[i:i+N]] for i in xrange(0, M*N, N)]

def get_neighbors(p):
    global M, N

    for dp in ((-1, 0), (1, 0), (0, -1), (0, 1)):
        np = (p[0] + dp[0], p[1] + dp[1])

        if np[0] < 0 or np[0] >= M or np[1] < 0 or np[1] >= N:
            continue

        yield np

def bfs(start, goal):
    global opened, M, N
    dist = [[-1]*N for i in xrange(M)]

    dist[start[0]][start[1]] = 0
    q = [start]

    while q:
        nq = []

        for p in q:
            if p == goal:
                return dist[goal[0]][goal[1]]

            for np in get_neighbors(p):
                if dist[np[0]][np[1]] == -1 and opened[np[0]][np[1]]:
                    dist[np[0]][np[1]] = dist[p[0]][p[1]] + 1
                    nq.append(np)

        q = nq

    return -1

dist = bfs(start, goal)

if dist == -1:
    print 'STAY WHERE YOU ARE'
else:
    print dist-1
